src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs

```diff
@@ -0,0 +1,560 @@
+            {
+                // `wasi:http/outgoing-handler` requires a non-empty authority,
+                // so we set one here:
+                if (scheme.Tag == ITypes.Scheme.HTTPS)
+                {
+                    authority = ":443";
+                }
+                else
+                {
+                    authority = ":80";
+                }
+            }
+            else
+            {
+                authority = uri.Authority;
+            }
+
+            var headers = new List<(string, byte[])>();
+            foreach (var pair in request.Headers)
+            {
+                foreach (var value in pair.Value)
+                {
+                    headers.Add((pair.Key, Encoding.UTF8.GetBytes(value)));
+                }
+            }
+
+            var outgoingRequest = new ITypes.OutgoingRequest(ITypes.Fields.FromList(headers));
+            outgoingRequest.SetScheme(scheme);
+            outgoingRequest.SetAuthority(authority);
+            outgoingRequest.SetPathWithQuery(uri.PathAndQuery);
+
+            var outgoingStream = new OutputStream(outgoingRequest.Body());
+
+            Func<Task<ITypes.IncomingResponse?>> sendContent = async () =>
+            {
+                await SendContentAsync(request.Content, outgoingStream).ConfigureAwait(false);
+                return null;
+            };
+
+            // Concurrently send the request and the content stream, allowing
+            // the server to start sending a response before it's received the
+            // entire request body.
+            var incomingResponse = (
+                await Task.WhenAll(
+                        new Task<ITypes.IncomingResponse?>[]
+                        {
+                            SendRequestAsync(outgoingRequest),
+                            sendContent()
+                        }
+                    )
+                    .ConfigureAwait(false)
+            )[0];
+
+            if (incomingResponse is null)
+            {
+                // Shouldn't be possible, since `SendRequestAsync` always
+                // returns a non-null value.
+                throw new Exception("unreachable code");
+            }
+
+            var response = new HttpResponseMessage((HttpStatusCode)incomingResponse.Status());
+            var responseHeaders = incomingResponse.Headers().Entries();
+            response.Content = new StreamContent(new InputStream(incomingResponse.Consume()));
+            foreach ((var key, var value) in responseHeaders)
+            {
+                var valueString = Encoding.UTF8.GetString(value);
+                // TODO: What's the right way to determine which headers go
+                // where?  Is there an easier way than explicitly enumerating
+                // all the possibilities in
+                // https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httpcontentheaders?view=net-8.0#properties
+                // ?
+                if (
```

Comment:
`src\libraries\System.Net.Http\src\System\Net\Http\Headers\HeaderDescriptor.cs` seems to be used in other HTTP clients